File: 1/2020.10.27/6道股票问题/123 最多二次交易/java/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int dp10 = 0, dp20 = 0;
        int dp11 = Integer.MIN_VALUE, dp21 = Integer.MIN_VALUE;
        for (int price : prices) {
            dp20 = Math.max(dp20, dp21 + price);
            dp21 = Math.max(dp21, dp10 - price);
            dp10 = Math.max(dp10, dp11 + price);
            dp11 = Math.max(dp11, -price);
        }
        return dp20;
    }
}

File: 1/2020.10.27/6道股票问题/123 最多二次交易/python/solution.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp10 = dp20 = 0
        dp11 = dp21 = float('-inf')
        for price in prices:
            dp20 = max(dp20, dp21 + price)
            dp21 = max(dp21, dp10 - price)
            dp10 = max(dp10, dp11 + price)
            dp11 = max(dp11, -price)
        return dp20
